Userland/Utilities/readelf.cpp

```diff
@@ -579,5 +582,68 @@ int main(int argc, char** argv)
         printf("\n");
     }
 
+    if (display_hardening) {
+        printf("Security Hardening:\n");
+        printf("RELRO        Stack Canary NX           PIE    Symbols\n");
+
+        // FIXME: Add support for full/partial RELRO detection using DT_BIND_NOW
+        bool relro = false;
+        interpreter_image.for_each_program_header([&relro](const ELF::Image::ProgramHeader& program_header) {
+            if (program_header.type() == PT_GNU_RELRO) {
+                relro = true;
+                return IterationDecision::Break;
+            }
+            return IterationDecision::Continue;
+        });
+
+        if (relro)
+            printf("\033[0;32m%-12s\033[0m ", "RELRO");
+        else
+            printf("\033[0;31m%-12s\033[0m ", "No RELRO");
+
+        bool canary = false;
+        interpreter_image.for_each_symbol([&canary](const ELF::Image::Symbol& sym) {
+            if (sym.name() == "__stack_chk_fail" || sym.name() == "__intel_security_cookie") {
+                canary = true;
+                return IterationDecision::Break;
+            }
+            return IterationDecision::Continue;
+        });
+
+        if (canary)
+            printf("\033[0;32m%-12s\033[0m ", "Canary found");
+        else
+            printf("\033[0;31m%-12s\033[0m ", "No canary");
+
+        bool nx = false;
+        interpreter_image.for_each_program_header([&nx](const ELF::Image::ProgramHeader& program_header) {
+            if (program_header.type() == PT_GNU_STACK) {
+                if (program_header.flags() == 0x07)
```

Comment:
Checking for RWX/RWE (`0x07`) is consistent with checksec. There's an [open issue](https://github.com/slimm609/checksec.sh/issues/132) discussing the shortcomings of this approach.

I also need to rewatch the [OS hacking: Stronger W^X memory protection (like PaX MPROTECT) video](https://www.youtube.com/watch?v=7e3qH0HZ1Eo) to check how R^X was implemented on SerenityOS given that we don't have a `GNU_STACK` header.

Reply:
Stacks are simply never executable on SerenityOS.

Reply:
I've modified this to check only if the stack is executable, regardless of whether it is read/write.

```
if (program_header.flags() & PF_X)
```

This is in line with the ELF validation in `LibELF` which calls shenanigans if the stack is executable.

https://github.com/SerenityOS/serenity/blob/0d8c9024eec12d57270cca1cf6bfc7d03c0ca229/Userland/Libraries/LibELF/Validation.cpp#L293-L298

I've also opened a PR #5867 to build with `--noexecstack` so all ELF files are build with the `GNU_STACK` ELF program header.